Dev

SaifAbuR3d opened the pull request.

SaifAbuR3d committed:
Add 'Filter Flights' feature

SaifAbuR3d committed:
read and write from csv and txt

SaifAbuR3d committed:
read passenger file and create instances with passenger repository

SaifAbuR3d committed:
view passenger bookings feature

SaifAbuR3d committed:
book a flight (passenger can write in booking.txt file) feature

SaifAbuR3d committed:
test write in files

SaifAbuR3d committed:
remove a booking feature
rewrite the bookings.txt file with updated list doesn't contain the deleted booking.

SaifAbuR3d committed:
implement model validations to ImportFlightsFromCsv
return a list of errors

SaifAbuR3d committed:
manager can reflect metadata about type
get the details about the type, constraints that should be applied on each field

SaifAbuR3d committed:
Add validation check for flights
each flight must have 3 different classes (Economy, Business, First Class)

SaifAbuR3d committed:
the flight is identified by (id,flightClass)

SaifAbuR3d committed:
The passenger can update the flight class

SaifAbuR3d committed:
add the User Interface

SaifAbuR3d committed:
Add Search feature to User Interface
edited BookingSearchCriteria for more convenient search

SaifAbuR3d committed:
move FlightSearchCriteria into seperate file
remove unnecessary using directives

SaifAbuR3d committed:
check if the booking already exists before adding/removing a booking

SaifAbuR3d committed:
make consistent null checking with pattern matching
and follow Properties naming conventions,
make static fields in Program readonly

SaifAbuR3d committed:
make Exit choice work properly

SaifAbuR3d committed:
Add Show Validation Details Choice for Manager menu

SaifAbuR3d committed:
Add consistency check for input flights cvs files

flights with the same FlightId must have identical data but with different Flight Classes/Prices.

Refactoring: Move input validation into a seperate class for better separating of concerns

SaifAbuR3d committed:
Remove unnecessary comments and done TODO

SaifAbuR3d committed:
Remove Add Flight method

SaifAbuR3d committed:
Add Tests project files

SaifAbuR3d committed:
Introduce BookingStatus as return type for easy testing

SaifAbuR3d committed:
Make the classes and interfaces public

SaifAbuR3d committed:
make the services classes public and make return type of .UpdateBookingClass method as BookingOperationStatus enum
to test the class easier.

SaifAbuR3d committed:
Add aditional check before modifying the flight
check if the passenger has really booked this flight
and fix the printed message.

ibrahim-masri commented on ATB.BL/DataAccess/CsvUtility.cs:
```diff
@@ -0,0 +1,115 @@
+        {
+            var csvConfig = new CsvConfiguration(CultureInfo.CurrentCulture)
+            {
+                HasHeaderRecord = false
```
should this be a fixed config? if not, how will you make it configurable?

ibrahim-masri commented on ATB.BL/DataAccess/CsvUtility.cs:
```diff
@@ -0,0 +1,115 @@
+                    var departureDate = DateTime.Parse(csvReader.GetField(4));
+                    var departureAirport = csvReader.GetField(5);
+                    var arrivalAirport = csvReader.GetField(6);
+                    var fClass = Enum.Parse<FlightClass>(csvReader.GetField(7), true);
```
how liberal are we with the input here? if the user has the id after the departure country, what would happen?

ibrahim-masri commented on ATB.BL/DataAccess/CsvUtility.cs:
```diff
@@ -0,0 +1,115 @@
+
+namespace ATB.DataAccess
+{
+    internal static class CsvUtility
```
why is this class internal? from where can I access it?

ibrahim-masri commented on ATB.BL/DataAccess/CsvUtility.cs:
```diff
@@ -0,0 +1,115 @@
+            return flights;
+        }
+
+        public static IEnumerable<Flight> ReadFlightsFromCsv(string csvFilePath)
```
is there shared logic between these two? if so how can I encapsulate it in one place?

ibrahim-masri commented on ATB.BL/DataAccess/CsvUtility.cs:
```diff
@@ -0,0 +1,115 @@
+            }
+            try 
+            {
+                using (StreamWriter writer = new StreamWriter(flightsFilePath, true, Encoding.UTF8))  // append the flights to file
```
what's the point of this using?

ibrahim-masri commented on ATB.BL/DataAccess/CsvUtility.cs:
```diff
@@ -0,0 +1,115 @@
+using CsvHelper;
+using CsvHelper.Configuration;
+
+namespace ATB.DataAccess
```
is this data access logic?

ibrahim-masri commented on ATB.BL/DataAccess/FileBookingRepository.cs:
```diff
@@ -0,0 +1,119 @@
+{
+    internal class FileBookingRepository : IBookingRepository
+    {
+        private string bookingsFilePath = "files/Bookings.txt";
```
same configuration question here, I want to make this configurable, how would I do it?

ibrahim-masri commented on ATB.BL/DataAccess/FileBookingRepository.cs:
```diff
@@ -0,0 +1,119 @@
+                File.AppendAllLines(bookingsFilePath, new[] { bookingLine });
+                Console.WriteLine("Booking added successfully!");
+            }
+            catch (IOException ex)
```
is this the best place for the catch? if not where should it be?

ibrahim-masri commented on ATB.BL/DataAccess/FileBookingRepository.cs:
```diff
@@ -0,0 +1,119 @@
+                {
+                    string[] values = line.Split(',');
+
+                    if (values.Length >= 3 && int.TryParse(values[0], out int passengerId))
```
are we tolerating bad data?

ibrahim-masri commented on ATB.BL/DataAccess/FileBookingRepository.cs:
```diff
@@ -0,0 +1,119 @@
+                    if (values.Length >= 3 && int.TryParse(values[0], out int passengerId))
+                    {
+                        int flightId = int.Parse(values[1]);
+                        Passenger? passenger = passengerRepository.GetPassenger(passengerId);
```
var maybe?

ibrahim-masri commented on ATB.BL/DataAccess/FileBookingRepository.cs:
```diff
@@ -0,0 +1,119 @@
+                    {
+                        int flightId = int.Parse(values[1]);
+                        Passenger? passenger = passengerRepository.GetPassenger(passengerId);
+                        FlightClass flightClass = Enum.Parse<FlightClass>(values[2], true); // true to ignore case
```
why are we ignoring case?

ibrahim-masri commented on ATB.BL/DataAccess/FileBookingRepository.cs:
```diff
@@ -0,0 +1,119 @@
+        }
+
+
+        public void AddBooking(Passenger passenger, Flight flight, FlightClass flightClass)
```
maybe moving these two overloaded methods together? the file is a bit difficult to navigate through?

ibrahim-masri commented on ATB.BL/DataAccess/FileBookingRepository.cs:
```diff
@@ -0,0 +1,119 @@
+                {
+                    foreach (Booking booking in bookings)
+                    {
+                        string line = $"{booking.Passenger.PassengerId},{booking.Flight.FlightId},{booking.FClass}"; 
```
this class is doing too much already, can we move this to a separate class? if so what would be the best option?

ibrahim-masri commented on ATB.BL/DataAccess/FileBookingRepository.cs:
```diff
@@ -0,0 +1,119 @@
+        public void RemoveBooking(Booking booking)
+        {
+            // Contains All Bookings Except For (booking)
+            IEnumerable<Booking> updatedList = GetAllBookings().Where(_booking => !_booking.Equals(booking));
```
this seems super expensive, what would be a more efficient way?

ibrahim-masri commented on ATB.BL/DataAccess/FileFlightRepository.cs:
```diff
@@ -0,0 +1,77 @@
+        public void AddAllFlights(IEnumerable<Flight> flights)
+        {
+
+            CsvUtility.AppendFlightsToCsv(flightsFilePath, flights);
```
it seem different repositories are having different layers of abstraction, when are we reading immediately from files and when are we using the CSVHelper?

ibrahim-masri commented on ATB.BL/DataAccess/FileFlightRepository.cs:
```diff
@@ -0,0 +1,77 @@
+
+        public FileFlightRepository()
+        {
+            InitializeFlightDictionary();
```
is there a point in making this a lazy load?

ibrahim-masri commented on ATB.BL/DataAccess/FileFlightRepository.cs:
```diff
@@ -0,0 +1,77 @@
+            return flightDictionary.Values;
+        }
+
+        public Flight? GetFlight(int flightId, FlightClass flightClass) // returns null if the Flight is not in the dictionary
```
is this the best place to be put this documentation?

ibrahim-masri commented on ATB.BL/DataAccess/FileFlightRepository.cs:
```diff
@@ -0,0 +1,77 @@
+
+            try
+            {
+                string[] lines = File.ReadAllLines(flightsFilePath);
```
validation?

ibrahim-masri commented on ATB.BL/DataAccess/FileFlightRepository.cs:
```diff
@@ -0,0 +1,77 @@
+                            FClass = fClass
+                        };
+                    }
+                }
```
duplicated code?

ibrahim-masri commented on ATB.BL/DataAccess/IBookingRepository.cs:
```diff
@@ -0,0 +1,15 @@
+        void UpdateBookingClass(Booking booking, FlightClass newClass);
+        void RemoveBooking(Booking booking);
+        void RemoveBooking(Passenger passenger, Flight flight, FlightClass flightClass);
+        IEnumerable<Booking> GetAllBookings();
```
what's the point in returning IEnumerable?

ibrahim-masri commented on ATB.BL/DataAccess/InputFlightDataValidation.cs:
```diff
@@ -0,0 +1,99 @@
+    {
+        FlightValidationResult validationResult = new();
+
+        if (!int.TryParse(csvReader.GetField(0), out int flightId))
```
are violating OCP?

ibrahim-masri commented on ATB.BL/DataAccess/InputFlightDataValidation.cs:
```diff
@@ -0,0 +1,99 @@
+    {
+
+        var allFlights = new FileFlightRepository().GetAllFlights();
+        var ConsistentFlights = new List<Flight>();
```
nit: naming

ibrahim-masri commented on ATB.BL/DataAccess/InputFlightDataValidation.cs:
```diff
@@ -0,0 +1,99 @@
+            }
+        }
+
+        return ConsistentFlights;
```
what's the point in returning the flight even if they are not consistent?

ibrahim-masri commented on ATB.BL/DataAccess/InputFlightDataValidation.cs:
```diff
@@ -0,0 +1,99 @@
+        var ConsistentFlights = new List<Flight>();
+        foreach (var flightGroup in flightGroups)
+        {
+            if (allFlights.Any(flight => flight.FlightId == flightGroup.Key))
```
multiple enumeration?

ibrahim-masri commented on ATB.BL/DataAccess/InputFlightDataValidation.cs:
```diff
@@ -0,0 +1,99 @@
+        return validationResult;
+    }
+
+    public static List<Flight> GetConsistentFlights(Dictionary<int, List<Flight>> flightGroups)
```
what is a consistent flight?

ibrahim-masri commented on ATB.BL/Entities/Booking.cs:
```diff
@@ -1,15 +1,39 @@
-        public Flight flight { get; set; }
-        public Passenger passenger { get; set; }
-        public Booking(Flight? _flight, Passenger? _passenger )
+        public Passenger Passenger { get; set; }
+        public Flight Flight { get; set; }
+        public FlightClass FClass { get; set; }
+        public  DateTime ReservationDate { get; set; }
+        public Booking(Flight flight, Passenger passenger, FlightClass fClass)
         {
```
why is this data in the main branch?

ibrahim-masri commented on ATB.BL/Entities/Booking.cs:
```diff
@@ -1,15 +1,39 @@
 
+        public override int GetHashCode()
+        {
+            throw new NotImplementedException();
```
why override ride it then?

ibrahim-masri commented on ATB.BL/Entities/Flight.cs:
```diff
@@ -1,28 +1,83 @@
-        FirstClass
-    }
-    internal readonly struct Flight
+    public readonly struct Flight
```
why a struct?

ibrahim-masri commented on ATB.BL/Entities/Flight.cs:
```diff
@@ -1,28 +1,83 @@
-        public string ArrivalAirport { get; }
-
-        public Flight(decimal price, string departureCounrty, string destinationCountry, DateOnly departureDate, string departureAirport, string arrivalAirport)
+        [Required(ErrorMessage = "Flight ID is required")]
```
where are these attributes used?

ibrahim-masri commented on ATB.BL/Entities/Flight.cs:
```diff
@@ -1,28 +1,83 @@
+        [DataType(DataType.Currency)]
+        [Required(ErrorMessage = "Price is required")]
+        [Range(1.0, 20000.0, ErrorMessage =
+            "Price Must be a nonnegative decimal in range (1.0 to 20000.0).")]
```
these messages make the code a bit messy, is there a better place to put this?

ibrahim-masri commented on ATB.BL/Entities/Manger.cs:
```diff
@@ -1,6 +1,8 @@
 {
-    internal class Manger : User
+
+    [Obsolete]
```
why keep it then :sweat_smile: ?

ibrahim-masri commented on ATB.sln:
```diff
@@ -15,6 +17,10 @@ Global
 		{45C18DEC-A517-43C8-B89C-53FBB568AA07}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{45C18DEC-A517-43C8-B89C-53FBB568AA07}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{45C18DEC-A517-43C8-B89C-53FBB568AA07}.Release|Any CPU.Build.0 = Release|Any CPU
+		{A57671B2-4423-4856-90B1-D748280D8640}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{A57671B2-4423-4856-90B1-D748280D8640}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{A57671B2-4423-4856-90B1-D748280D8640}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{A57671B2-4423-4856-90B1-D748280D8640}.Release|Any CPU.Build.0 = Release|Any CPU
```
what do these do?

ibrahim-masri commented on ATB.BL/Services/FlightService.cs:
```diff
@@ -0,0 +1,46 @@
+        { 
+            var allFlights = GetAllFlights();
+            return allFlights
+                .Where(flight =>
```
OCP?

ibrahim-masri commented on ATB.BL/Helpers/UserInputHelper.cs:
```diff
@@ -0,0 +1,103 @@
+            decimal result = 0;
+            bool isValidInput = false;
+
+            while (!isValidInput)
```
is there a documentation for how the user input looks like?

ibrahim-masri commented on ATB.BL/Helpers/ValidationDetailsHelper.cs:
```diff
@@ -0,0 +1,43 @@
+    {
+        internal static void DisplayValidationDetails(Type EntityType)
+        {
+            var properties = EntityType.GetProperties();
```
fast member?

ibrahim-masri commented on ATB.BL/Presentation/Program.cs:
```diff
@@ -1,9 +1,443 @@
+                Console.WriteLine("Welcome to Airport Ticket Booking App!");
+                Console.WriteLine("-------------------------------------");
+                Console.WriteLine("1. Passenger");
+                Console.WriteLine("2. Manager");
```
how do we apply more security here?

ibrahim-masri commented on ATB.BL/Presentation/Program.cs:
```diff
@@ -1,9 +1,443 @@
+        }
+
+
+        private static void ImportFlightsFromCSV()
```
what is this doing in Program.cs

SaifAbuR3d committed:
remove unused method .ReadFlightsFromCsv
reorder methods in FileBookingRepository for better readability

SaifAbuR3d committed:
Encapsulate the logic of creation CsvReader objects into a Factory class, with more configurations options.

SaifAbuR3d committed:
encapsulate the parsing logic into a separate class
following OCP and SRP

SaifAbuR3d committed:
Encapsulate the creation of the flight line string into a separate method for better code organization.

SaifAbuR3d committed:
Encapsulate the logic of parsing a string[] and creating a Flight, aligning with SRP and OCP.

SaifAbuR3d committed:
fix naming violation

SaifAbuR3d committed:
xml documentation for .GetFlight better than random comment

SaifAbuR3d committed:
use var to simplify code

SaifAbuR3d committed:
delete obsolete class 'Manager'

SaifAbuR3d committed:
improve readability of .FilterBookings method

SaifAbuR3d committed:
improve readability of .FilterFlights method

SaifAbuR3d committed:
Print the error message for required attributes in .DisplayValidationDetails

SaifAbuR3d committed:
Propagate handling exeptions from BookingFileRepository to services layer
-which provide more context to handle them appropriately. (logging, printing to screen, ...)

SaifAbuR3d committed:
Propagate handling exeptions from FileFlightRepository and CsvUtility to services layer (where possible)
which provide more context to handle them appropriately.
